### DISCUSSION TITLE
Postgresql dialect `OnConflictClause` unexpected behavior when providing `schema.Index` as constraint

### DISCUSSION THREAD

**Discussion by kasbaker** [sqlalchemy/sqlalchemy#9022] [Usage Questions] [answered]
I am working on an upsert operation where I want to use a `schema.Index` as the `conflict_target` https://www.postgresql.org/docs/current/sql-insert.html

While working on this I found that `sqlalchemy.dialects.postgresql.insert.on_conflict_do_update` has some unexpected behavior when you provide an `Index` instead of a `Constraint`.

My schema has a polymorphic table that looks something like this:
```python
from sqlalchemy.sql.functions import coalesce

from sqlalchemy import Column, Integer, Index, String
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class MyTable(Base):
    __tablename__ = "my_table"

    id = Column(Integer, primary_key=True)
    name = Column(String)
    col1 = Column(String, nullable=True)
    col2 = Column(String, nullable=True)
    ix_upsert = Index(
        "ix_upsert",
        name,
        coalesce(col1, col2),
        unique=True,
    )
```

When I do an upsert like this
```python
from sqlalchemy.dialects.postgresql import insert
data = [
    {"name": "foo", "col1": "qux", "col2": None},
    {"name": "bar","col1": None, "col2": "quux"}
]
column_names = data[0].keys()

stmt = insert(MyTable).values(data)
upsert = stmt.on_conflict_do_update(
    constraint=MyTable.ix_upsert,
    set_={
        key: val for key, val in stmt.excluded.items() if key in column_names
    },
)
````
I get an error like this:
`sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedObject) constraint "ix_upsert" for table "my_table" does not exist`

I was able to fix the error by passing `MyTable.ix_upsert.expressions` as `index_elements` to the upsert like this: 
```python
upsert = stmt.on_conflict_do_update(
    index_elements=MyTable.ix_upsert.expressions,
    set_={
        key: val for key, val in stmt.excluded.items() if key in column_names
    },
)
```
I figured out how to fix it when looking through the source code and I found some unexpected behavior in `sqlalchemy.dialects.postgresql.dml.OnConflictClause`: 
https://github.com/sqlalchemy/sqlalchemy/blob/b973cbd8939f2cc0e29c668fffd507958c3e455a/lib/sqlalchemy/dialects/postgresql/dml.py#L208

This line stringifies the constraint: https://github.com/sqlalchemy/sqlalchemy/blob/b973cbd8939f2cc0e29c668fffd507958c3e455a/lib/sqlalchemy/dialects/postgresql/dml.py#L218

So that this block never executes:
https://github.com/sqlalchemy/sqlalchemy/blob/b973cbd8939f2cc0e29c668fffd507958c3e455a/lib/sqlalchemy/dialects/postgresql/dml.py#L230

By removing `schema.Index` from the type check on this earlier line, the clause would behave more as expected:
https://github.com/sqlalchemy/sqlalchemy/blob/b973cbd8939f2cc0e29c668fffd507958c3e455a/lib/sqlalchemy/dialects/postgresql/dml.py#L216

The variable name `constraint` doesn't quite make sense here, so perhaps `conflict_target` would be better since it is consistent with the Postgres documentation.

**Comment by zzzeek** [comment, marked as answer]
this is a bug but has never been reported; the PG docs, as far as I can tell, do not clarify that "constraint_name" at https://www.postgresql.org/docs/current/sql-insert.html refers to a UNIQUE or EXCLUDE constraint name, but **not** the name of a UNIQUE INDEX , even though that is also most certainly a "constraint" in its nature, just not in PG's internal terminology.

This stackoverflow answer seems to clarify:  https://dba.stackexchange.com/a/139773

bigger bug is of course how is it that we dont have a round trip test with this particular combination.


removing Index from that particular check, since we can't use index names here at all, would be the fix

```diff
@@ -188,7 +189,7 @@ class OnConflictClause(ClauseElement):
         if constraint is not None:
             if not isinstance(constraint, util.string_types) and isinstance(
                 constraint,
-                (schema.Index, schema.Constraint, ext.ExcludeConstraint),
+                (schema.Constraint, ext.ExcludeConstraint),
             ):
                 constraint = getattr(constraint, "name") or constraint
 
```

this is #9023

**Reply by zzzeek** [reply]
Oh, I see, what's unexpected here is that you are passing Index to the argument called "constraint", which, per PG itself, is not what it considers to be a "constraint" :)   .  What we are then doing is accepting it as a constraint, rather than a. not accepting it or b. expanding it out.    We will do b. but that's why this use case didnt seem to get noticed (dont know how the code got written, though).

**Reply by kasbaker** [reply]
Thanks @zzzeek. I didn't know whether to list it as a bug because as you said I am passing an Index to an argument called "constraint". Expanding it out seems safer, though it does complicate things a bit.
